numpy/linalg/linalg.py

```diff
@@ -918,14 +918,14 @@ def qr(a, mode='reduced'):
     # determine optimal lwork
     lwork = 1
     work = zeros((lwork,), t)
-    results = lapack_routine(m, mc, mn, q, m, tau, work, -1, 0)
+    results = lapack_routine(m, mc, mn, q, max(1, m), tau, work, -1, 0)
     if results['info'] != 0:
         raise LinAlgError('%s returns %d' % (routine_name, results['info']))
 
     # compute q
-    lwork = int(abs(work[0]))
+    lwork = max(1, n, int(abs(work[0])))
```

Comment:
So, looking at the source code - `work[0]` is populated as `N*NB`, and `NB = 32`` - so strictly speaking the `n` is not needed here, Arguably, lapack ought to be setting `work[0]` to `max(1, N*NB)`, so I'd consider this a bug in lapack.

Either way, what you have here is definitely correct, and it doesn't really do any harm to be extra safe.